Tests/FluentAssertions.Equivalency.Specs/DataTableSpecs.cs

```diff
@@ -32,6 +33,112 @@ public void When_data_tables_are_both_null_equivalence_test_should_succeed()
+    public void When_row_match_mode_is_invalid_it_should_fail()
+    {
+        // Arrange
+        var typedDataSet = CreateDummyDataSet<TypedDataSetSubclass>();
+
+        var subject = typedDataSet.ToUntypedDataSet().Tables["TypedDataTable1"];
+        var expectation = typedDataSet.ToUntypedDataSet().Tables["TypedDataTable1"];
+
+        // Act
+        Action action = () => subject.Should().BeEquivalentTo(expectation, options => options.UsingRowMatchMode((RowMatchMode)2));
+
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "Unknown RowMatchMode *when trying to compare *");
+    }
+
+    [Theory]
+    [MemberData(nameof(EmptyPrimaryKeys))]
+    public void When_row_match_mode_is_primary_key_without_primary_key_it_should_fail(DataColumn[] emptyPrimaryKey)
+    {
+        // Arrange
+        var typedDataSet = CreateDummyDataSet<TypedDataSetSubclass>(includeRelation: false);
+
+        var subject = typedDataSet.ToUntypedDataSet().Tables["TypedDataTable1"];
+        var expectation = typedDataSet.ToUntypedDataSet().Tables["TypedDataTable1"];
+
+        subject.PrimaryKey = emptyPrimaryKey;
+
+        // Act
+        Action action = () =>
+            subject.Should().BeEquivalentTo(expectation, options => options.UsingRowMatchMode(RowMatchMode.PrimaryKey));
+
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Table *containing *does not have a primary key. RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    public static IEnumerable<object[]> EmptyPrimaryKeys => new List<object[]>
+    {
+        new object[] { null },
+        new object[] { new DataColumn[] { } }
+    };
+
+    [Fact]
+    public void When_primary_key_types_do_not_match_it_should_throw()
+    {
+        // Arrange
+        var typedDataSetSubject = CreateDummyDataSet<TypedDataSetSubclass>(includeDummyData: false, includeRelation: false);
+        var typedDataSetExpectation = new TypedDataSetSubclass(typedDataSetSubject);
+
+        var subject = typedDataSetSubject.ToUntypedDataSet().Tables["TypedDataTable1"];
+        var expectation = typedDataSetExpectation.ToUntypedDataSet().Tables["TypedDataTable1"];
+
+        subject.PrimaryKey[0].DataType = typeof(long);
+        subject.Rows.Add(1L);
+        subject.AcceptChanges();
+        expectation.Rows.Add(1);
+        expectation.AcceptChanges();
+
+        // Act
+        Action action = () =>
+            subject.Should().BeEquivalentTo(expectation, options => options.UsingRowMatchMode(RowMatchMode.PrimaryKey));
+
+        // Assert
+        action.Should().Throw<XunitException>().WithMessage(
+            "*Subject and expectation primary keys of table containing *do not have the same schema and cannot be compared. " +
+            "RowMatchMode.PrimaryKey cannot be applied.*");
+    }
+
+    [Theory]
+    [InlineData(1)]
+    [InlineData(3)]
+    public void When_primary_key_of_rows_differ_it_should_fail(int nbrRowsToEdit)
```

Comment:
I would prefer duplicating this test to avoid conditional logic like `for` and `if`.